python-checks/src/main/java/org/sonar/python/checks/regex/GroupReplacementCheck.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks.regex;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonarsource.analyzer.commons.regex.RegexParseResult;
+import org.sonarsource.analyzer.commons.regex.ast.CapturingGroupTree;
+import org.sonarsource.analyzer.commons.regex.ast.RegexBaseVisitor;
+
+import static org.sonar.python.tree.TreeUtils.nthArgumentOrKeyword;
+
+
+@Rule(key = "S6328")
+public class GroupReplacementCheck extends AbstractRegexCheck {
+
+  private static final String MESSAGE = "Referencing non-existing group%s: %s.";
+  private static final Pattern REFERENCE_PATTERN = Pattern.compile("\\$(\\d+)|\\$\\{(\\d+)}|\\\\(\\d+)");
+
+  @Override
+  protected Map<String, Integer> lookedUpFunctions() {
+    Map<String, Integer> result = new HashMap<>();
+    result.put("re.sub", 4);
+    return result;
+  }
+
+  @Override
+  public void checkRegex(RegexParseResult regexParseResult, CallExpression regexFunctionCall) {
+    GroupFinder groupFinder = new GroupFinder();
+    groupFinder.visit(regexParseResult);
+    checkReplacement(regexFunctionCall, groupFinder.groups);
+  }
+
+  private void checkReplacement(CallExpression tree, Set<CapturingGroupTree> groups) {
+    RegularArgument regArg = nthArgumentOrKeyword(1, "replacement", tree.arguments());
+    if(regArg == null){
+      return;
+    }
```

Comment:
```suggestion
    if(regArg == null) return;
```
Very minor (my personal preference)